db/version_edit.h

```diff
@@ -660,6 +673,22 @@ class VersionEdit {
 
   const char* DecodeNewFile4From(Slice* input);
 
+  void EncodeFileBoundaries(std::string* dst, const FileMetaData& meta,
+                            size_t ts_sz) const {
+    if (ts_sz == 0 || meta.user_defined_timestamps_persisted) {
+      PutLengthPrefixedSlice(dst, meta.smallest.Encode());
+      PutLengthPrefixedSlice(dst, meta.largest.Encode());
+      return;
+    }
+    std::string smallest_buf;
```

Comment:
I know that this is not performance critical code, but it bothers me somewhat that there's an extra key copy to do this. Not really because of this code but I expect something similar to happen in code that is more performance critical. I believe it's not simple because the timestamp is between the user key and the seqno+kind, but there could be a version of StripTimestampFromInternalKey that includes the length prefix. Or something like SliceParts could be used (but SliceParts does not encapsulate the sequence of Slices, just provides a view into an existing array of Slices). Something to think about.

Reply:
I appreciate a lot all the performance related suggestions! You are right that even this path is not a hot one, similar ideas are used in more performance critical places. Also this is a great idea! It looks like there is an implementation `PutLengthPrefixedSliceParts` that is ready to use to help save this one copy. I will work on a follow up to update this and places with similar logic to use an in place stripping version of function `StripTimestampFromInternalKey` that returns a SliceParts, and encoding with `PutLengthPrefixedSliceParts` to save this extra copy.